Contribution guide updates (#2028) (#2038)
DataBiosphere/azul: CONTRIBUTING.rst

* For Python we use PEP8 with E722 disabled (type too general in except clause)
  and the maximum line length set to 120 characters.

* For prose (documentation, messages, comments) wrap lines at the word boundary
  closest to or at, but not beyond, column 79. The first column is column 0.

* For code, we keep the trimmed line length under 81. A trimmed line is a line

>   in the source with leading and trailing whitespace removed. This means a line

Reviewer: I just realized, this may be redundant since the presence of trailing whitespace is automatically a flake8 error. Unless the intent here is only to exclude the newline character.

* For Python we use PEP8 with E722 disabled (type too general in except clause)
  and the maximum line length set to 120 characters.

* For prose (documentation, messages, comments) wrap lines at the word boundary
  closest to or at, but not beyond, column 79. The first column is column 0.

* For code, we keep the trimmed line length under 81. A trimmed line is a line
  in the source with leading and trailing whitespace removed. This means a line
  may be indented by 40 characters and contain 80 characters after that. This
  rule is designed to keep code readable without forcing excessive wrapping for

>   more deeply nested control flow constructs.

Reviewer 2: We still have the absolute width rule of 120, right? I have this plugged into PyCharm from the formatting file I believe.

  rule is designed to keep code readable without forcing excessive wrapping for
  more deeply nested control flow constructs.

* We prefer single quoted string literals. We used to use double quotes in JSON
  literals but that convention is now deprecated and all new string literals are

>   single quoted except as noted below.

Reviewer 2: When writing new code around old JSON literals do we match style, or covert all to single quotes?

    self.call_me(foo, bar,
                 x=1, y=2)

>   The one exception to this rule are logging method invocations and calls to

Reviewer 3: ```suggestion
  The only exceptions to this rule are logging method invocations and calls to
```

  methods. The distinction between instance and class methods is driven by
  higher order concerns than the one about whether a method's body currently
  references self or not.

* When catching expected exceptions, especially for EAFP, we minimize the body
  of the try block::

    d = make_my_dict()
    try:
        x = d['x']
    except:
        <do stuff without x>
    else:
        <do stuff with x>

  This is not a mere cosmetic convention, it affects program correctness. If the
  call to ``make_my_dict`` were done inside the ``try`` block, a KeyError raised

>   by it would be conflated with the one raised d['x']. The latter is expected,

Reviewer 3: ```suggestion
  by it would be conflated with the one raised by d['x']. The latter is expected,
```

  refactorings or minor unrelated changes. The title of merge commit in this
  case usually matches that of the main commit.

* Github lets any user with write access resolve comments to changes in a PR. We
  aren't that permissive. When the reviewer makes a comment, either requesting
  a change or asking a question, the author addresses the comment by either

>   - making the requested changes and reacting to the comment with thumbs-up

Reviewer 4: ```suggestion
  - making the requested changes and reacting to the comment with thumbs-up 👍 
```

>     - making the requested changes and reacting to the comment with thumbs-up
> 
>     - or answering the question in a reply to the comment.

Reviewer 5: Nitpick - dedent to prevent rendering as `<blockquote>`
```suggestion
  - making the requested changes and reacting to the comment with thumbs-up

  - or answering the question in a reply to the comment.
```

  refactorings or minor unrelated changes. The title of merge commit in this
  case usually matches that of the main commit.

* Github lets any user with write access resolve comments to changes in a PR. We
  aren't that permissive. When the reviewer makes a comment, either requesting
  a change or asking a question, the author addresses the comment by either

  - making the requested changes and reacting to the comment with thumbs-up

  - or answering the question in a reply to the comment.

>   In either case, only the reviewer resolves the comment. This is to ensure that

Reviewer 2: This doesn't address the case where the reviewer's suggestions were incorrect or couldn't be applied without caveats.
